pkg/apis/seedmanagement/validation/managedseed_test.go

```diff
@@ -0,0 +1,477 @@
+				})),
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeRequired),
+					"Field": Equal("metadata.namespace"),
+				})),
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
+					"Field": Equal("metadata.namespace"),
+				})),
+			))
+		})
+
+		It("should forbid namespace different from garden", func() {
+			managedSeed.Namespace = "foo"
+
+			errorList := ValidateManagedSeed(managedSeed)
+
+			Expect(errorList).To(ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
+					"Field": Equal("metadata.namespace"),
+				})),
+			))
+		})
+
+		It("should forbid empty shoot name", func() {
+			managedSeed.Spec.Shoot.Name = ""
+
+			errorList := ValidateManagedSeed(managedSeed)
+
+			Expect(errorList).To(ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeRequired),
+					"Field": Equal("spec.shoot.name"),
+				})),
+			))
+		})
+
+		It("should forbid both seed template and gardenlet", func() {
+			managedSeed.Spec.Gardenlet = &seedmanagement.Gardenlet{}
+
+			errorList := ValidateManagedSeed(managedSeed)
+
+			Expect(errorList).To(ConsistOf(
+				PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
+					"Field": Equal("spec"),
+				})),
+			))
+		})
+
+		Context("seed template", func() {
+			It("should allow valid resources", func() {
+				errorList := ValidateManagedSeed(managedSeed)
+
+				Expect(errorList).To(HaveLen(0))
+			})
+
+			It("should forbid empty or invalid fields in seed template", func() {
+				seedCopy := seed.DeepCopy()
+				seedCopy.Spec.Provider.Type = ""
+				managedSeed.Spec.SeedTemplate.Spec = seedCopy.Spec
+
+				errorList := ValidateManagedSeed(managedSeed)
+
+				Expect(errorList).To(ConsistOf(
+					PointTo(MatchFields(IgnoreExtras, Fields{
+						"Type":  Equal(field.ErrorTypeRequired),
+						"Field": Equal("spec.seedTemplate.spec.provider.type"),
+					})),
+				))
+			})
+		})
+
+		Context("gardenlet", func() {
+			var (
+				seedx *gardencorev1beta1.Seed
+				err   error
+			)
+
+			BeforeEach(func() {
+				seedx, err = helper.ConvertSeedExternal(seed)
+				Expect(err).NotTo(HaveOccurred())
+
+				managedSeed.Spec.SeedTemplate = nil
+				managedSeed.Spec.Gardenlet = &seedmanagement.Gardenlet{
+					Deployment: &seedmanagement.GardenletDeployment{
+						Image: &seedmanagement.Image{
+							PullPolicy: pullPolicyPtr(corev1.PullIfNotPresent),
+						},
+					},
+					Config:          gardenletConfiguration(seedx, nil),
+					Bootstrap:       bootstrapPtr(seedmanagement.BootstrapToken),
+					MergeWithParent: pointer.BoolPtr(true),
+				}
+			})
+
+			It("should allow valid resources", func() {
+				errorList := ValidateManagedSeed(managedSeed)
+
+				Expect(errorList).To(HaveLen(0))
+			})
+
+			It("should forbid empty or invalid fields in gardenlet", func() {
+				seedxCopy := seedx.DeepCopy()
```

Comment:
Just FYI: You don't need to `DeepCopy` here. `BeforeEach` is executed before each `It` and will reset the variables.

Reply:
OK, you are right, will remove it.